### ISSUE TITLE
connectInfiniteHits has undefined hits when updating state

### ISSUE THREAD

**Issue by Haroenv** [algolia/react-instantsearch#883]
(copied from https://github.com/algolia/instantsearch.js/issues/2663 by @udewilde)

Hi,

I have implemented connectInfiniteHits [react-infinite-scroller](https://github.com/CassetteRocks/react-infinite-scroller), and all is working almost perfectly.
But whenever the state of the React Component is modified (not related to the infiniteHits), the connector is re-rendered with value `undefined` for the `hits` prop, which breaks the page.

Inside the `connectInfinitHits` connector :
It looks like `refine` is firing whenever the state of the component changes, leading to the issue.
`this.previousPage` and `hits` are `undefined`, so following logic loop is ignored, and `this._allResults` is `undefined` as well.

```js
if (page === 0) { 
  this._allResults = hits; 
} else if (page > this.previousPage) { 
  this._allResults = [].concat(_toConsumableArray(this._allResults), _toConsumableArray(hits)); 
} else if (page < this.previousPage) { 
  this._allResults = hits; 
}
```
Am I doing something wrong?

**Comment by samouss** [comment]
Hi @udewilde, I don't manage to reproduce the error. Could you share with us a little reproduction of what you are trying to do? It will be a lot easier to dig the issue. We provide template in order to avoid you the boilerplate part. You can find it on [CodeSandbox](https://codesandbox.io/s/github/algolia/instantsearch-templates/tree/master/src/React%20InstantSearch). Thanks!

**Comment by udewilde** [comment]
Sure, sorry for this.
Here is the [JSFiddle](https://jsfiddle.net/udewilde/v0ezceen/2/)
The equivalent of clicking the button throws
```
Uncaught (in promise) TypeError: Cannot read property 'map' of undefined
```

and if repeating the click:
```
ReactReconciler.js:60 Uncaught (in promise) 
TypeError: Cannot read Uncaught (in promise) 
TypeError: Cannot read property 'getHostNode' of null
```

**Comment by samouss** [comment]
The jsFilddle don't work but I made a small example in [CodeSandox](https://codesandbox.io/s/6x54538vv3) from it. I saw that in your fiddle you are calling the connector on each render. You should avoid to do that and create the `NeverEndingResult` component outside of your `App` component. Check the example, now when you click on the `button` it should not trigger any error.

**Comment by udewilde** [comment]
Ok you make it look too simple now!
Thanks a lot, this was the issue.

**Comment by samouss** [comment]
You're welcome! I will close this issue, feel free to reopen it if needed 😉

**Comment by oyeanuj** [comment]
@Haroenv @samouss I came here with a similar error though I am not sure if it is the same root cause. As I go from a search page of one type of object, to another type of object, the `infiniteHits` breaks. 

It is hard to reproduce in a CodeSandbox publicly but here is the code that will help elucidate the structure:

```jsx
// Each SearchResults page (navigated via nav component below)

<PageContainer>
  <Nav />
  <Configure />  //<-------- From Algolia
  <SearchResultsGrid />
</PageContainer>

// Depending on the url, the configure properties will change
```
And the `SearchResultsPage` looks like this:
```jsx
@withRouter
@connectInfiniteHits
@connectStateResults
@connect(gridSelector)
export default class SearchResultsGrid extends Component {
  ...
}
```

I get the following error:
<img width="772" alt="screen shot 2018-01-25 at 8 57 01 pm" src="https://user-images.githubusercontent.com/9633371/35425991-81f1e0f6-0214-11e8-9dbf-5430561a9106.png">


On investigating further, I noticed the following things: 
1. On navigation, even though new `SearchResultsGrid` component is being created, notice how initially the `hits` are undefined and `hasMore` is true. This is even before the query is seen to be made on the console. Should `hits` ever be undefined and shouldn't `hasMore` be false until results are fetched?

2. I also noticed in the `searchState` prop from `connectStateResult` that the page number and the  `configure` `filter` prop hadn't reset on the page throwing the error. On a subsequent re-render the `configure` prop first gets dropped and then changes but the page number doesn't change. You can see that in the screenshot below.

<img width="766" alt="screen shot 2018-01-25 at 9 26 20 pm" src="https://user-images.githubusercontent.com/9633371/35426254-77fe7558-0216-11e8-8766-71e871d0d163.png">


My guess is what I noticed above in (2) is leading to the errors but I am not sure what might be the reason for that behavior in the first place? Thoughts?

**Comment by oyeanuj** [comment]
@Haroenv thanks for the quick response! I did a little more debugging and noticed that if I a passed a `key` to `InstantSearch` component that would make it re-render everytime `Configure` props changed, then things work well. 

So, it seems like there is some weirdness with Configure where the change in props to Configure don't propagate well to `InstantSearch` or the actual search and hence the stale `page` and `configure.filter` props above in the screenshots.

I'll also try to recreate the above in a CodeSandbox.

**Comment by samouss** [comment]
@oyeanuj We released a version with a fix on the `connectInfiniteHits` connector. It might solve your issue. The fix has been released on the version [v5.0.1](https://github.com/algolia/react-instantsearch/blob/master/CHANGELOG.md#501-2018-03-12).

**Comment by samouss** [comment]
I'm closing this issue, since we released a fix that should solve the issue. If you still have it please open a new one. Thanks!
